game-core/src/main/java/games/strategy/engine/data/DefaultAttachment.java

```diff
@@ -152,11 +152,10 @@ public boolean equals(final Object obj) {
       return false;
     }
     final DefaultAttachment other = (DefaultAttachment) obj;
-    if (m_attachedTo == null) {
-      if (other.m_attachedTo != null) {
-        return false;
-      }
-    } else if (!m_attachedTo.toString().equals(other.m_attachedTo.toString())) {
+    if (m_attachedTo != null && other.m_attachedTo != null) {
+      return m_attachedTo.toString().equals(other.m_attachedTo.toString());
+    }
+    if (m_attachedTo != null || other.m_attachedTo != null) {
       return false;
     }
```

Comment:
I don't think so, your proposal doesn't really return true depending on the attachments name...
This equals method really isn't pretty and doesn't 100% align with hashCode.
So this method needs an overhaul anyways if you needed something to make ou start working on it ^^

Reply:
But it looks like the new code is now short-circuiting

```java
Objects.equals(m_name, other.m_name) || this.toString().equals(other.toString())
```

if

```java
m_attachedTo != null && other.m_attachedTo != null
```

Previously if `m_attachedTo.toString().equals(other.m_attachedTo.toString())` was `true`, it would fall through to the `return` on old line 162.  Now, it's just returning `true` without testing `m_name`, right?

Reply:
Sounds reasonable, but it's currently almost 3 am here, so I don't think I'm in the right state to make a qualified statement here 😅
If you're convinced I can just put your code in instead without any further thoughts.

Reply:
I'm always convinced by unit tests. :smile:  I'll try to write some before I go offline later this evening to verify there's no change in behavior.

Reply:
Ok then.
In this case I'd appreciate if you could put this change in a follow-up PR ^^

Reply:
Sure, if you want.  But I figure I can get them written and this issue sorted out one way or the other by the time you wake up. :smile:  If for some reason that's not the case, then I suppose we can do this one change in a separate PR.

Reply:
The following characterization tests provide 100% line and branch coverage over the original `equals()` implementation:

<details>
<summary>Unit tests</summary>

```diff
diff --git a/game-core/src/test/java/games/strategy/engine/data/DefaultAttachmentTest.java b/game-core/src/test/java/games/strategy/engine/data/DefaultAttachmentTest.java
index 514e60803..d668b63e3 100644
--- a/game-core/src/test/java/games/strategy/engine/data/DefaultAttachmentTest.java
+++ b/game-core/src/test/java/games/strategy/engine/data/DefaultAttachmentTest.java
@@ -8,6 +8,9 @@ import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
 
+import java.util.Collections;
+import java.util.Map;
+
 import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
@@ -15,6 +18,141 @@ import org.mockito.Mock;
 import org.mockito.junit.jupiter.MockitoExtension;
 
 final class DefaultAttachmentTest {
+  @ExtendWith(MockitoExtension.class)
+  @Nested
+  final class EqualsTest {
+    private static final String NAME = "attachment";
+    private static final String OTHER_NAME = "otherAttachment";
+
+    @Mock
+    private Attachable attachable;
+    @Mock
+    private Attachable otherAttachable;
+    private final GameData gameData = new GameData();
+
+    private class TestDefaultAttachment extends DefaultAttachment {
+      private static final long serialVersionUID = 1L;
+
+      private final String stringValue;
+
+      TestDefaultAttachment(
+          final String name,
+          final Attachable attachable,
+          final GameData gameData,
+          final String stringValue) {
+        super(name, attachable, gameData);
+
+        this.stringValue = stringValue;
+      }
+
+      @Override
+      public void validate(final GameData data) {}
+
+      @Override
+      public Map<String, MutableProperty<?>> getPropertyMap() {
+        return Collections.emptyMap();
+      }
+
+      @Override
+      public String toString() {
+        return stringValue;
+      }
+    }
+
+    private DefaultAttachment givenAttachment(final String name, final Attachable attachable) {
+      return givenAttachment(name, attachable, name);
+    }
+
+    private DefaultAttachment givenAttachment(
+        final String name,
+        final Attachable attachable,
+        final String stringValue) {
+      return new TestDefaultAttachment(name, attachable, gameData, stringValue);
+    }
+
+    @Test
+    void shouldReturnTrueWhenOtherIsSameInstance() {
+      final DefaultAttachment attachment = givenAttachment(NAME, attachable);
+      final DefaultAttachment otherAttachment = attachment;
+
+      assertThat(attachment.equals(otherAttachment), is(true));
+    }
+
+    @Test
+    void shouldReturnFalseWhenOtherIsNull() {
+      final DefaultAttachment attachment = givenAttachment(NAME, attachable);
+      final DefaultAttachment otherAttachment = null;
+
+      assertThat(attachment.equals(otherAttachment), is(false));
+    }
+
+    @Test
+    void shouldReturnFalseWhenOtherIsDifferentClass() {
+      final DefaultAttachment attachment = givenAttachment(NAME, attachable);
+      final DefaultAttachment otherAttachment = new TestDefaultAttachment(NAME, attachable, gameData, NAME) {
+        private static final long serialVersionUID = 1L;
+      };
+
+      assertThat(attachment.equals(otherAttachment), is(false));
+    }
+
+    @Test
+    void shouldReturnFalseWhenThisAttachableIsNotNullAndOtherAttachableIsNull() {
+      final DefaultAttachment attachment = givenAttachment(NAME, attachable);
+      final DefaultAttachment otherAttachment = givenAttachment(NAME, null);
+
+      assertThat(attachment.equals(otherAttachment), is(false));
+    }
+
+    @Test
+    void shouldReturnFalseWhenThisAttachableIsNullAndOtherAttachableIsNotNull() {
+      final DefaultAttachment attachment = givenAttachment(NAME, null);
+      final DefaultAttachment otherAttachment = givenAttachment(NAME, otherAttachable);
+
+      assertThat(attachment.equals(otherAttachment), is(false));
+    }
+
+    @Test
+    void shouldReturnTrueWhenThisAttachableIsNullAndOtherAttachableIsNullAndNameEqual() {
+      final DefaultAttachment attachment = givenAttachment(NAME, null);
+      final DefaultAttachment otherAttachment = givenAttachment(NAME, null);
+
+      assertThat(attachment.equals(otherAttachment), is(true));
+    }
+
+    @Test
+    void shouldReturnFalseWhenAttachableNotEqual() {
+      final DefaultAttachment attachment = givenAttachment(NAME, attachable);
+      final DefaultAttachment otherAttachment = givenAttachment(NAME, otherAttachable);
+
+      assertThat(attachment.equals(otherAttachment), is(false));
+    }
+
+    @Test
+    void shouldReturnFalseWhenAttachableEqualAndNameNotEqual() {
+      final DefaultAttachment attachment = givenAttachment(NAME, attachable);
+      final DefaultAttachment otherAttachment = givenAttachment(OTHER_NAME, attachable);
+
+      assertThat(attachment.equals(otherAttachment), is(false));
+    }
+
+    @Test
+    void shouldReturnTrueWhenAttachableEqualAndNameEqual() {
+      final DefaultAttachment attachment = givenAttachment(NAME, attachable);
+      final DefaultAttachment otherAttachment = givenAttachment(NAME, attachable);
+
+      assertThat(attachment.equals(otherAttachment), is(true));
+    }
+
+    @Test
+    void shouldReturnTrueWhenAttachableEqualAndNameNotEqualAndToStringEqual() {
+      final DefaultAttachment attachment = givenAttachment(NAME, attachable, "stringValue");
+      final DefaultAttachment otherAttachment = givenAttachment(OTHER_NAME, attachable, "stringValue");
+
+      assertThat(attachment.equals(otherAttachment), is(true));
+    }
+  }
+
   @ExtendWith(MockitoExtension.class)
   @Nested
   final class GetAttachmentTest {
```

</details>

The test `shouldReturnFalseWhenThisAttachableIsNotNullAndOtherAttachableIsNull()` errors with an NPE using the existing implementation (:+1: for LGTM).

Using the implementation from this PR fixes the NPE, but now the test `shouldReturnFalseWhenAttachableEqualAndNameNotEqual()` fails.  That seems to confirm my short-circuit theory above.

All tests pass using the following implementation (diff is relative to the current state of this branch):

```diff
diff --git a/game-core/src/main/java/games/strategy/engine/data/DefaultAttachment.java b/game-core/src/main/java/games/strategy/engine/data/DefaultAttachment.java
index c7f9c2086..2e679a9dd 100644
--- a/game-core/src/main/java/games/strategy/engine/data/DefaultAttachment.java
+++ b/game-core/src/main/java/games/strategy/engine/data/DefaultAttachment.java
@@ -152,12 +152,7 @@ public abstract class DefaultAttachment extends GameDataComponent implements IAt
       return false;
     }
     final DefaultAttachment other = (DefaultAttachment) obj;
-    if (m_attachedTo != null && other.m_attachedTo != null) {
-      return m_attachedTo.toString().equals(other.m_attachedTo.toString());
-    }
-    if (m_attachedTo != null || other.m_attachedTo != null) {
-      return false;
-    }
-    return Objects.equals(m_name, other.m_name) || this.toString().equals(other.toString());
+    return Objects.equals(Objects.toString(m_attachedTo, null), Objects.toString(other.m_attachedTo, null))
+        && (Objects.equals(m_name, other.m_name) || this.toString().equals(other.toString()));
   }
 }
```